### PR TITLE
Enable driver<->app communication with ClusterIP

### PR DESCRIPTION
**Pull request by hrosenbe**
This pull request has the following changes:
* When running the workload drivers and applications on the same cluster, it is now possible to have the Nginx service created as a ClusterIP service.  In this case the communication between the drivers and the applications will only use cluster-internal IP addresses.
* The _useLoadBalancer_ parameter has been removed and replaced with the _appIngressMethod_ parameter.  The appIngressMethod parameter can take the values loadbalancer, nodeport, or clusterip.
    * The User's guide has been updated to reflect this change.
    * A check and message has been added to runWeathervane.pl to notify users for the change.
* In order to remove the need for cluster-external IP addresses for communication between the run harness and the workload controller, all such communication now uses _kubectl exec_ to run http commands inside the wkldcontroller container (using the httpie command).  
* Weathervane can now run on Minikube.  The documentation was been updated with Minikube specific information.
* As a workaround for clusters with DNS issues, the IP address of a Cassandra node is always used as the listen address, rather than the hostname.  In addition, on cassandra-0 it is added to the list of seeds.
* The resource requests have been removed from the AuctionDataManager pod so that it can be scheduled when running with small nodes.
* The wait time for all services starting has been increased to accommodate runs with large numbers of instances on slow nodes.
* Services are now started in batches of 10 with a 60 second inter-batch delay.  Previously all services were started simultaneously  This delay fixes issues with inability to start all instances on large runs.

### CODE REVIEW

File: doc/userDocs/usersGuide.md [hunk 1]
```diff
@@ -1137,7 +1190,7 @@ Configuration parameters not mentioned here have an empty string as the default
 | `runStrategy`               | `findMaxSingleRun`            |
 | `numAppInstances`           | `1`                           |
 | `kubeconfigFile`            | `~/.kube/config`              |
-| `useLoadBalancer`           | `TRUE`                        |
+| `appIngressMethod`          | `loadbalancer                |
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: missing closing back quote?

File: dockerImages/runharness/Dockerfile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Copyright 2017-2019 VMware, Inc.
 # SPDX-License-Identifier: BSD-2-Clause
-FROM namespace/weathervane-centos7:version
+FROM hrosenbe/weathervane-centos7:intraCluster
```
Comment (bhoflich) [inline dockerImages/runharness/Dockerfile]: Use namespace instead of hrosenbe?
What does intraCluster do?

File: runHarness/Utils.pm [hunk 3]
```diff
@@ -255,6 +255,43 @@ sub callBooleanMethodOnObjectsParallel2 {
...
+sub callBooleanMethodOnObjectsParallel2BatchDelay {
+	my ( $method, $objectsRef, $param1, $param2, $batchSize, $delaySec ) = @_;
+	my $console_logger = get_logger("Console");
+	my @pids;
+	my $pid;
+	my $objectNum = 0;
+
+	foreach my $object (@$objectsRef) {
+		if ($objectNum == $batchSize) {
+			sleep($delaySec);
```
Comment (bhoflich) [inline runHarness/Utils.pm]: I suggest printing a log message (perhaps debug) when a sleep occurs.
Here in 2 param and in 3 param also.

File: configFiles/kubernetes/auctionDataManager.yaml [hunk 4]
```diff
@@ -70,10 +70,6 @@ spec:
             - /loadData.pl
           initialDelaySeconds: 15
           periodSeconds: 10
-        resources:
-          requests:
-            cpu: 1
-            memory: 8Gi
```
Comment (bhoflich) [inline configFiles/kubernetes/auctionDataManager.yaml]: Will this change impact db load time?

File: doc/userDocs/usersGuide.md [hunk 5]
```diff
@@ -141,7 +140,32 @@ The image below shows the lines that need to be edited in the configuration file
...
+      clusters, you can choose either:
+        - `loadbalancer`:  Traffic from the workload drivers to the applications will
+          use the external IP address of a Kubernetes LoadBalancer service. Your cluster must 
+          support provisioning external IP addresses for LoadBalancer services. 
+        - `nodeport`:  Traffic from the workload drivers to the applications will
+          use the external IP addresses of the Kubernetes Nodes via a NodePort service. Your cluster must 
+          support exposing the node IP addresses externally using NodePort services.
+    - If the workload drivers and the applications are running on the same Kubernetes 
+      cluster, you can choose among:
+        - `none`: Traffic from the workload drivers to the applications will use a
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: Should 'none' be 'clusterip' ?

File: doc/userDocs/usersGuide.md [hunk 6]
```diff
@@ -141,7 +140,32 @@ The image below shows the lines that need to be edited in the configuration file
...
+          use the external IP address of a Kubernetes LoadBalancer service. This 
+          may be desirable even with a single cluster in order to load-test the complete networking 
+          stack of your cluster. Your cluster must 
+          support provisioning external IP addresses for LoadBalancer services. 
+        - `nodeport`: Traffic from the workload drivers to the applications will
+          use the external IP addresses of the Kubernetes Nodes via a NodePort service. This 
+          may be desirable even with a single cluster in order to load-test the complete networking 
+          stack of your cluster. Your cluster must 
+          support exposing the node IP addresses externally using NodePort services.
+
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: There is a lot of duplication here.   Can the clusterip case and "desirable even with a single cluster" text be factored out to allow only 1 loadbalance and 1 nodeport section?

File: doc/userDocs/usersGuide.md [hunk 7]
```diff
@@ -141,7 +140,32 @@ The image below shows the lines that need to be edited in the configuration file
 2. Update *kubeconfigFile* to the full path to your kubeconfig file.  The paths will be identical if using a single cluster for both the app and driver.
 3. Update *kubeconfigContext* parameters with the name of the context for your cluster from the kubeconfig file, such as `kubernetes-admin@kubernetes`.
     - Set an empty string (`""`) to use the current-context defined in the kubeconfigFile.
-4. Set *useLoadBalancer* to `true` if your cluster supports LoadBalancer services. Otherwise, set it to `false` to use a NodePort for ingress.
+4. Choose a value for *appIngressMethod* that reflects the method you want to use
+  for communication between the workload drivers and the applications.
+    - If the workload drivers and applications are running on different Kubernetes 
+      clusters, you can choose either:
+        - `loadbalancer`:  Traffic from the workload drivers to the applications will
+          use the external IP address of a Kubernetes LoadBalancer service. Your cluster must
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: 'Your cluster"  -- does this mean the driver cluster or the application cluster, or both?

File: doc/userDocs/usersGuide.md [hunk 8]
```diff
@@ -141,7 +140,32 @@ The image below shows the lines that need to be edited in the configuration file
...
-4. Set *useLoadBalancer* to `true` if your cluster supports LoadBalancer services. Otherwise, set it to `false` to use a NodePort for ingress.
+4. Choose a value for *appIngressMethod* that reflects the method you want to use
+  for communication between the workload drivers and the applications.
+    - If the workload drivers and applications are running on different Kubernetes 
+      clusters, you can choose either:
+        - `loadbalancer`:  Traffic from the workload drivers to the applications will
+          use the external IP address of a Kubernetes LoadBalancer service. Your cluster must 
+          support provisioning external IP addresses for LoadBalancer services. 
+        - `nodeport`:  Traffic from the workload drivers to the applications will
+          use the external IP addresses of the Kubernetes Nodes via a NodePort service. Your cluster must
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: "Your cluster" again.

File: doc/userDocs/usersGuide.md [hunk 9]
```diff
@@ -206,16 +230,39 @@ Handling clusters whose credential expire is discussed [below](#expire).
...
+The Weathervane workload driver interacts with the Weathervane application through a Kubernetes service fronting the application's web-server 
+tier. Weathervane 
+allows the user to select the type of Kubernetes service to be used.  Your selection will depend 
+on the types of services supported by your cluster and the goals of your tests.  
+
+Weathervane supports the use of 
+[LoadBalancer, NodePort, or ClusterIP services](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types).   You 
+indicate the type of service you want to use by setting the *appIngressMethod* parameter 
+to `loadbalancer`, `nodeport`, or `clusterip`.  The default value for this parameter 
+is `loadbalancer`.
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: Maybe add a note that for vanilla / custom created clusters (non cloud), that loadbalancer will likely not apply and this will need to be changed from the default value?

File: doc/userDocs/usersGuide.md [hunk 10]
```diff
@@ -1356,4 +1409,79 @@ As a result, when using node labels it is recommended that you label all of your
...
+running.
+
+#### Weathervane Configuration
+
+In order to run Weathervane on Minikube, you need to use the following configuration 
+parameter values:
+
+| Parameter | Value |
+|-----------|-------|
+| appIngressMethod | clusterIp |
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: should be 'clusterip' ? (lower case i)

File: doc/userDocs/usersGuide.md (hunk 1 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: Yes.  Will fix in next commit.

File: dockerImages/runharness/Dockerfile (hunk 2 above)
Comment (hrosenbe) [reply dockerImages/runharness/Dockerfile]: Good catch.  This is something that must have crept in because I aborted a build of the runharness container. intraCluster was what I was using for the version number (in version.txt).

File: runHarness/Utils.pm (hunk 3 above)
Comment (hrosenbe) [reply runHarness/Utils.pm]: Added a debug log message for now.

File: configFiles/kubernetes/auctionDataManager.yaml (hunk 4 above)
Comment (hrosenbe) [reply configFiles/kubernetes/auctionDataManager.yaml]: No.  The requests are only used for admission control.  For micro the 1 CPU request is more than the CPU requests of the non-dataServices pods, and so I was running into situations where it couldn't start the data manager pod even though there would be enough space for all of the instances once the data was prepared.

File: doc/userDocs/usersGuide.md (hunk 5 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: yes.  Good catch.  I initially had a slightly different interpretation of the appIngressMethod parameter, and was using _none_ rather than _clusterip_.

File: doc/userDocs/usersGuide.md (hunk 6 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: Agreed.  I have refactored this to remove the duplication.

File: doc/userDocs/usersGuide.md (hunk 7 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: The application cluster.  I will clarify

File: doc/userDocs/usersGuide.md (hunk 8 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: will clarify

File: doc/userDocs/usersGuide.md (hunk 9 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: agreed.  i will add this.

File: doc/userDocs/usersGuide.md (hunk 10 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: yes.  I just pushed a commit with fixes for these comments.

File: doc/userDocs/usersGuide.md (hunk 9 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: added

File: doc/userDocs/usersGuide.md (hunk 8 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: done

File: doc/userDocs/usersGuide.md [hunk 11]
```diff
@@ -141,29 +141,21 @@ The image below shows the lines that need to be edited in the configuration file
...
-        - `none`: Traffic from the workload drivers to the applications will use a 
-          cluster internal address provisioned by a Kubernetes ClusterIP service.  
-          This method will work on all clusters.
-        - `loadbalancer`: Traffic from the workload drivers to the applications will
+  for communication between the workload drivers and the applications. You can choose 
+  from among:
+    - `clusterip`: Traffic from the workload drivers to the applications will use a 
+          cluster internal address provisioned by a Kubernetes ClusterIP service.  This 
+          choice will only work if the workload drivers and applications are running 
+          on the same cluster.
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: I forgot to ask, what happens if you specify clusterip with multiple clusters?
Is there a check, or does it just run into errors, or?

Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: Good question.  There should be a check but there is not.  Currently it will just not work, and I don't know what errors it will give (as I only have a single cluster).  I will add a check.

File: doc/userDocs/usersGuide.md [hunk 12]
```diff
@@ -241,7 +233,9 @@ Weathervane supports the use of
 [LoadBalancer, NodePort, or ClusterIP services](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types).   You 
 indicate the type of service you want to use by setting the *appIngressMethod* parameter 
 to `loadbalancer`, `nodeport`, or `clusterip`.  The default value for this parameter 
-is `loadbalancer`.
+is `loadbalancer`.  Node that cluster created manually using kubeadm will likely not 
+support LoadBalancer services.  In than case you will need to use either NodePort or
```
Comment (bhoflich) [inline doc/userDocs/usersGuide.md]: than -> that

File: runHarness/AppInstance/AuctionAppInstance.pm [hunk 13]
```diff
@@ -111,21 +110,25 @@ override 'checkConfig' => sub {
...
 		return 0;
 	}
 
-	my $validAppInstanceSizes = $WeathervaneTypes::appInstanceSizes->{"auction"};
-	if ( $self->getParamValue('appInstanceSize') ~~ @$validAppInstanceSizes ) {
-		$console_logger->error("Workload $workloadNum, AppInstance $appInstanceNum: The AppInstance size for the Auction workload must be one of: @$validAppInstanceSizes");
-		return 0;
-	}
+    my $validAppInstanceSizes = $WeathervaneTypes::appInstanceSizes{"auction"};
+    if (!(grep { $self->getParamValue('configurationSize') eq $_ } @$validAppInstanceSizes)) {
```
Comment (RebeccaYo) [inline runHarness/AppInstance/AuctionAppInstance.pm]: Question: why did you change the text matching method from '~~' to grep?

File: doc/userDocs/usersGuide.md [hunk 14]
```diff
@@ -141,7 +140,24 @@ The image below shows the lines that need to be edited in the configuration file
 2. Update *kubeconfigFile* to the full path to your kubeconfig file.  The paths will be identical if using a single cluster for both the app and driver.
 3. Update *kubeconfigContext* parameters with the name of the context for your cluster from the kubeconfig file, such as `kubernetes-admin@kubernetes`.
     - Set an empty string (`""`) to use the current-context defined in the kubeconfigFile.
-4. Set *useLoadBalancer* to `true` if your cluster supports LoadBalancer services. Otherwise, set it to `false` to use a NodePort for ingress.
+4. Choose a value for *appIngressMethod* that reflects the method you want to use
+  for communication between the workload drivers and the applications. You can choose 
+  from among:
+    - `clusterip`: Traffic from the workload drivers to the applications will use a
```
Comment (RebeccaYo) [inline doc/userDocs/usersGuide.md]: I notice later in the doc you adopt the ordering "loadbalancer", "nodeport" and "clusterip". I like having the default listed first and cluster IP last as it requires only one cluster and doesn't stress the entire network stack, so could you put clusterip last.

File: doc/userDocs/usersGuide.md [hunk 15]
```diff
@@ -1356,4 +1403,79 @@ As a result, when using node labels it is recommended that you label all of your
...
+
+The command to start Minikube with 3 cpus and 8Gb is :
+* `minikube start --cpus=3 --memory=8g`
+
+Note that even with this cluster configuration, Minikube may have trouble running Weathervane 
+depending on how much memory is configured on the laptop and what else is 
+running.
+
+#### Weathervane Configuration
+
```
Comment (RebeccaYo) [inline doc/userDocs/usersGuide.md]: To run on Minikube, the user should still follow the Quickstart guide, correct? And the setup diverges when it comes to configuring Weathervane. For completeness, maybe add something like:
"To run Weathervane on Minikube, follow the Quickstart Setup steps Obtain Weathervane from GitHub and Create the Weathervane Docker Images. Then follow the instructions below to create the Weathervane Configuration file when running on Minikube."

File: doc/userDocs/usersGuide.md (hunk 12 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: Yes.  Should have been "Note that ..."

File: doc/userDocs/usersGuide.md (hunk 14 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: That ordering makes sense.  However the reason that I put clusterip first here is that it made it easier to explain that you might want to use loadbalancer or nodeport even on a single cluster.  

I will try putting clusterip last and rewording the others slightly.  Let me know what you think.

File: doc/userDocs/usersGuide.md (hunk 15 above)
Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: Good idea.  I will add this.

Comment (hrosenbe) [reply doc/userDocs/usersGuide.md]: I actually added this in an overview section so that we don't have to repeat it for evey cluster type we describe here.

File: runHarness/AppInstance/AuctionAppInstance.pm (hunk 13 above)
Comment (hrosenbe) [reply runHarness/AppInstance/AuctionAppInstance.pm]: I was having a strange issue with this test failing.  I thought that it might be due to some strange behavior of ~~, which is an experimental feature and is no longer really supported.  So I changed all of the ~~ matching in this file to use grep.  Then I realized that the test was using an otherwise unused parameter called _appInstanceSize_ rather than the correct _configurationSize_.  I must have used _appInstanceSize_ initially and then changed the name without deleting the parameter or changing it here.
